library/core/src/task/wake.rs

```diff
@@ -215,6 +283,125 @@ impl fmt::Debug for Context<'_> {
+    pub const fn from_waker(waker: &'a Waker) -> Self {
+        // SAFETY: LocalWaker is just Waker without thread safety
+        let local_waker = unsafe { transmute(waker) };
+        Self { waker: Some(waker), local_waker }
+    }
+
+    /// Create a ContextBuilder from a LocalWaker.
+    #[inline]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    pub const fn from_local_waker(local_waker: &'a LocalWaker) -> Self {
+        Self { local_waker, waker: None }
+    }
+
+    /// This field is used to set the value of the waker on `Context`.
+
+    #[inline]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    pub const fn waker(self, waker: &'a Waker) -> Self {
+        Self { waker: Some(waker), ..self }
+    }
+
+    /// This method is used to set the value for the local waker on `Context`.
+    #[inline]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    pub const fn local_waker(self, local_waker: &'a LocalWaker) -> Self {
+        Self { local_waker, ..self }
+    }
+
+    /// Builds the `Context`.
+    #[inline]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    pub const fn build(self) -> Context<'a> {
+        let ContextBuilder { waker, local_waker } = self;
+        Context { waker, local_waker, _marker: PhantomData, _marker2: PhantomData }
+    }
+}
+
+/// Construct a [`ContextBuilder`] from a [`Context`]. This is useful for
+/// overriding values from a context.
+///
+/// # Examples
+/// An example of a future that allows to set a [`Waker`] on Context if none was defined.
+/// This can be used to await futures that require a [`Waker`] even if the runtime does not
+/// support [`Waker`].
+/// ```rust
+/// #![feature(noop_waker, local_waker)]
+/// use std::task::{Waker, ContextBuilder};
+/// use std::future::{poll_fn, Future};
+/// use std::pin::pin;
+///
+/// async fn with_waker<F>(f: F, waker: &Waker) -> F::Output
+/// where
+///     F: Future
+/// {
+///     let mut f = pin!(f);
+///     poll_fn(move |cx| {
+///         let has_waker = cx.try_waker().is_some();
+///         if has_waker {
+///             return f.as_mut().poll(cx);
+///         }
+///
+///         let mut cx = ContextBuilder::from(cx)
+///             .waker(&waker)
+///             .build();
+///         f.as_mut().poll(&mut cx)
+///     }).await
+/// }
+///
+/// # async fn __() {
+/// with_waker(async { /* ... */ }, &Waker::noop()).await;
+/// # }
+/// ```
+#[unstable(feature = "local_waker", issue = "118959")]
+impl<'a> From<&mut Context<'a>> for ContextBuilder<'a> {
+    #[inline]
+    fn from(value: &mut Context<'a>) -> Self {
+        let Context { waker, local_waker, .. } = *value;
```

Comment:
`..` doesn't feel appropriate here, I'd rather see the fields explicitly written out so any additional fields get caught at compile time as needing changes here.

Reply:
Done